kasa/kasacamprotocol.py

```diff
@@ -0,0 +1,132 @@
+"""Protocol implementation for controlling TP-Link Kasa Cam devices."""
+
+import dataclasses
+import errno
+import logging
+from base64 import b64decode, b64encode
+from dataclasses import dataclass
+from pprint import pformat as pf
+from typing import Dict
+from urllib.parse import quote
+
+import httpx
+
+# When support for cpython older than 3.11 is dropped
+# async_timeout can be replaced with asyncio.timeout
+from async_timeout import timeout as asyncio_timeout
+
+from .credentials import Credentials
+from .exceptions import SmartDeviceException
+from .json import loads as json_loads
+from .protocol import TPLinkSmartHomeProtocol
+
+_LOGGER = logging.getLogger(__name__)
+_NO_RETRY_ERRORS = {errno.EHOSTDOWN, errno.EHOSTUNREACH, errno.ECONNREFUSED}
+
+
+@dataclass
+class KasaCamCredentials(Credentials):
+    """Adapter to store credentials in format expected by Kasa Cam devices."""
+
+    def __post_init__(self):
+        """Base64 encode the password as expected by the Kasa Cam protocol."""
+        self.password: str = b64encode(self.password.encode())
+
+    @classmethod
+    def from_instance(cls, instance):
+        """Create KasaCamCredentials from Credentials."""
+        return cls(**dataclasses.asdict(instance))
+
+
+class KasaCamProtocol(TPLinkSmartHomeProtocol):
+    """Implementation of the Kasa Cam protocol."""
+
+    DEFAULT_PORT = 10443
+
+    def __init__(self, *args, **kwargs) -> None:
+        self.credentials = kwargs.pop("credentials")
+        if not isinstance(self.credentials, KasaCamCredentials):
+            self.credentials = KasaCamCredentials.from_instance(self.credentials)
+
+        super().__init__(*args, **kwargs)
+
+        self.session: httpx.Client = None
```

Comment:
#509 adds a dependency for another http library (aiohttp) and we don't want to depend on two libraries for that. I'm wondering what are the pros & the cons of each of those libraries? Ping @sdb9696

Reply:
@rytilahti I started out with aiohttp but it appears they're opinionated about how they handle SSL in a manner that's contrary to most other libraries (requests, httpx, etc.) 

aiohttp requires building an SSL context to accept the self signed cert and I wasn't able to find a way to build a working context without having the self signed cert beforehand. "Disabling" ssl by passing "ssl=False" to the constructor also does not bypass this check, unfortunately. That's the only reason I have for using httpx, so if anyone is able to build a working context without needing the cert, I'd be happy to swap out to aiohttp

Reply:
Hi @bstrdsmkr, did you try creating a default ssl context and passing it in to aiohttp?
`sslcontext = ssl.create_default_context()`
Also the [aiohttp docs](https://docs.aiohttp.org/en/stable/client_advanced.html) seem to suggest you can just skip the strict checking by just catching the exceptions:
> If you need to skip both ssl related errors
```
try:
    await session.get('https://expired.badssl.com/')
except aiohttp.ClientSSLError as e:
    assert isinstance(e, ssl.SSLError)

try:
    await session.get('https://wrong.host.badssl.com/')
except aiohttp.ClientSSLError as e:
    assert isinstance(e, ssl.CertificateError)
```

Reply:
@sdb9696 I went on another dive and stumbled across a post that led to an answer to build a successful context for aiohttp. 
```shell
~ curl -v -k -s https://192.168.1.161:10443 2>&1 | grep "SSL connection using"                                                                                                  
* SSL connection using TLSv1.2 / AES256-GCM-SHA384
```
This gave me a working cipher so I built a context like this:
```python
ssl_context = ssl.create_default_context()
ssl_context.set_ciphers("AES256-GCM-SHA384")
ssl_context.check_hostname = False
ssl_context.verify_mode = ssl.CERT_NONE
self.session = aiohttp.ClientSession(
    base_url=f"https://{self.host}:{self.port}",
    connector=aiohttp.TCPConnector(ssl=ssl_context),
    auth=aiohttp.BasicAuth(
        self.credentials.username, self.credentials.password
    ),
)
```

While this works, ssl is outside of my wheelhouse and from what I've read this bypasses the cipher negotiation and only happens to work because my machine's openssl and the target device both allow that cipher. This feels fragile to me and my kung fu isn't strong enough to know how to build a more robust solution.

That said, I also looked into comparing the two libs and it looks to me like `httpx` is the "better buy" as it supports both sync and async mode, request tracing, and a few other nice-to-have things.

What are your thoughts on using `httpx` instead of `aiohttp`?

Reply:
Hi @bstrdsmkr, I saw similar information around httpx having some advantages and being a more complete library.  I've tested the klap changes with httpx and it seems to work fine and doesn't require too much work.  @rytilahti do you want me to change the klap stuff to work with httpx?  I can turn it around pretty quickly but I'd be very appreciative it if it doesn't slow down getting the PR in as I'm keen to close this one out.

Reply:
Hi, sorry for brevity, I'm busy at the moment. So my personal experiences are mostly with aiohttp, but that's only because it does not only implement the client but also the server side of HTTP. My understanding is that httpx is preferred for client applications, so let's go with it. Please do convert the klap PR to use httpx, @sdb9696.

@bstrdsmkr I'm wondering if it's really necessary to set the ciphers at all? Anyway, I think `CERT_NONE` should work just fine as to my understanding just instructs the context to skip the verification.

Reply:
> Please do convert the klap PR to use httpx, @sdb9696.

Ok that's done and pushed to https://github.com/python-kasa/python-kasa/pull/509